Media/RTShaderLib/SGXLib_NormalMap.glsl

```diff
@@ -41,20 +41,198 @@ void SGX_FetchNormal(in sampler2D s,
+						in vec2 texCoord,
+						in vec3 viewPos,
+						in vec2 scaleBias,
+						in mat3 TBN,
+						out vec2 newTexCoord)
+{
+	//Calculate eye direction
+	TBN = transpose(TBN);
+	vec3 eyeVec = mul(TBN, -viewPos);
+	eyeVec = normalize(eyeVec);
+	
+	//Simple parallax mapping
+	newTexCoord = texCoord;
+	float height = 1.0f - texture2D(normalHeightMap, newTexCoord).a;   
+	eyeVec.y = -eyeVec.y;
+    vec2 p = eyeVec.xy / eyeVec.z * (height * scaleBias.x);
+    newTexCoord = newTexCoord - p; 
+}
+
+void SGX_Generate_Parallax_Steep_Texcoord(in sampler2D normalHeightMap,
+						in vec2 texCoord,
+						in vec3 viewPos,
+						in vec2 scaleBias,
+						in mat3 TBN,
+						out vec2 newTexCoord)
+{
+	//Calculate eye direction
+	TBN = transpose(TBN);
+	vec3 eyeVec = mul(TBN, -viewPos);
+	eyeVec = normalize(eyeVec);
+	eyeVec.y = -eyeVec.y; //Inverse y
+	
+	//Steep parallax occlusion mapping
+	//Configure steep mapping layering.
+    const float numLayers = 32;	
+    float layerDepth = 1.0 / numLayers;	
+    float currentLayerDepth = 0.0;	
+    vec2 parallaxShift = (eyeVec.xy) * scaleBias.x; 
+    vec2 deltaTexCoords = parallaxShift / numLayers;
+	
+	newTexCoord = texCoord;	
+	float currentDepthMapValue = 1.0f - texture2D(normalHeightMap, newTexCoord).a;
+	
+	//Loop through layers and break early if match found.
+	for (int currentLayerId = 0; currentLayerId < (int)numLayers; currentLayerId++)
+	{
+		// shift texture coordinates along direction of P
+		newTexCoord -= deltaTexCoords;
+		
+		// get depthmap value at current texture coordinates
+		currentDepthMapValue = 1.0f - texture2D(normalHeightMap, newTexCoord).a;  
+		
+		//Break if layer height matched
+		if (currentLayerDepth > currentDepthMapValue)
+			break;
+		
+		// get depth of next layer
+		currentLayerDepth += layerDepth;  
+	}
+	
+	
+	// get texture coordinates before collision (reverse operations)
+	/*vec2 prevTexCoords = newTexCoord + deltaTexCoords;
+
+	// get depth after and before collision for linear interpolation
+	float afterDepth  = currentDepthMapValue - currentLayerDepth;
+	float beforeDepth = texture2D(normalHeightMap, prevTexCoords).a - currentLayerDepth + layerDepth;
+	 
+	// interpolation of texture coordinates
+	float weight = afterDepth / (afterDepth - beforeDepth);
+	newTexCoord = prevTexCoords * weight + newTexCoord * (1.0 - weight);*/	
+}
+
+void SGX_Generate_Parallax_Texcoord_Terrain(in sampler2D normalHeightMap,
+						in vec2 texCoord,
+						in vec3 viewPos,
+						in vec2 scaleBias,
+						in mat3 TBN,
+						out vec2 newTexCoord)
+{
+	//Calculate eye direction
+	TBN = transpose(TBN);
+	vec3 eyeVec = mul(TBN, -viewPos);
+	eyeVec = normalize(eyeVec);
+	
+	//Simple parallax mapping
+	newTexCoord = texCoord;
+	float height = 1.0f - texture2D(normalHeightMap, newTexCoord).a;   
+    vec2 p = eyeVec.xy * (height * scaleBias.x);
+    newTexCoord = newTexCoord - p;
+}
+
+void SGX_Generate_Parallax_Steep_Texcoord_Terrain(in sampler2D normalHeightMap,
```

Comment:
you could `#ifdef TERRAIN_SURFACE` the differences between this and the other version into one.
One will ever need only one variant of this per generated shader.

Reply:
I can't seem to get this to work properly - it just ignores the #ifdef within SGXLib_NormalMap.glsl?

Reply:
ah, sorry.. there is no `TERRAIN_SURFACE` define, I meant you define it.
But there already is `TERRAIN_STEEP_PARALLAX_MAPPING`, so you can use that..

Reply:
> ah, sorry.. there is no `TERRAIN_SURFACE` define, I meant you define it. But there already is `TERRAIN_STEEP_PARALLAX_MAPPING`, so you can use that..

Ah yes I see. I have sorted this now.